opal/mca/accelerator/accelerator.h

```diff
@@ -0,0 +1,475 @@
+
+/**
+ * Transfer types.
+ * UNSPEC - Not specified
+ * HTOH - Host to Host
+ * HTOD - Host to Device
+ * DTOH - Device to Host
+ * DTOD - Device to Device
+ */
+typedef enum {
+    MCA_ACCELERATOR_TRANSFER_UNSPEC = 0,
+    MCA_ACCELERATOR_TRANSFER_HTOH,
+    MCA_ACCELERATOR_TRANSFER_HTOD,
+    MCA_ACCELERATOR_TRANSFER_DTOH,
+    MCA_ACCELERATOR_TRANSFER_DTOD,
+} opal_accelerator_transfer_type_t;
+
+struct opal_accelerator_stream_t {
+    /* Stream object */
+    void *stream;
+};
+typedef struct opal_accelerator_stream_t opal_accelerator_stream_t;
+OBJ_CLASS_DECLARATION(opal_accelerator_stream_t);
+
+struct opal_accelerator_event_t {
+    /* Event object */
+    void *event;
+};
+typedef struct opal_accelerator_event_t opal_accelerator_event_t;
+OBJ_CLASS_DECLARATION(opal_accelerator_event_t);
+
+/**
+ * Check whether a pointer belongs to an accelerator or not.
+ * interfaces
+ *
+ * @param[IN] addr           Pointer to check
+ * @param[OUT] dev_id        Returns the device id against which the memory was allocated
+ * @param[OUT] flags         Set to 0 or an Accelerator flag to indicate additional
+ *                           information about the corresponding pointer
+ *
+ * @retval <0                An error has occurred.
+ * @retval 0                 The buffer does not belong to a managed buffer
+ *                           in device memory.
+ * @retval >0                The buffer belongs to a managed buffer in
+ *                           device memory.
+ */
+typedef int (*opal_accelerator_base_module_check_addr_fn_t)(
+    const void *addr, int *dev_id, uint64_t *flags);
+
+/**
+ * Creates a stream for asynchonous operations. This will also
+ * allocate the stream object.
+ *
+ * @param[IN] dev_id         Associated device for the stream
+ * @param[IN] stream         Stream to create
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_create_stream_fn_t)(
+    int dev_id, opal_accelerator_stream_t *stream);
+
+/**
+ * Destroys a stream created by opal_accelerator_base_module_create_stream_fn_t().
+ * This will also free the stream object.
+ *
+ * @param[IN] stream         Stream to destroy
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_destroy_stream_fn_t)(
+    opal_accelerator_stream_t *stream);
+
+/**
+ * Waits until all events on the stream are completed.
+ *
+ * @param[IN] stream         Stream to wait for
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_synchronize_stream_fn_t)(
+    opal_accelerator_stream_t *stream);
+
+/**
+ * Creates an event. You must allocate memory for event prior to
+ * this call.
+ *
+ * @param[IN] event          Event to create
+ *
+ * @return                   OPAL_SUCCESS or error status on failure.
+ */
+typedef int (*opal_accelerator_base_module_create_event_fn_t)(
```

Comment:
For habana you also need `device_id` when you create an event.

```
synStatus SYN_API_CALL synEventCreate( synEventHandle*       pEventHandler,
                                       const synDeviceId     deviceId,
                                       const uint32_t        flags );
```

Reply:
The API needs to make a decision and stick with it.  Either there's one device associated with a process and none of the calls require a device id (other than maybe a get/set_device_id) or they all need a device id.  I don't have strong preferences, but the argument that the underlying API has a device id doesn't make sense; instead it should be an argument about the framework's usage model.

Reply:
We decided to avoid calls with device_ids. I removed device_id from most of the API's other than the get device id. If a new component comes along with a different requirement of device_ids that cannot be adapted to the API comes along, then they can propose some changes and we can have a v2 of the API.